File: quick_sort/quick-sort.go
package main

import "fmt"

func QuickSort(in []int) []int {
	return quicksort(in, 0, len(in)-1)
}

func quicksort(in []int, low, high int) []int {
	//
	if low < high {
		partitionIdx := partition(in, low, high)
		quicksort(in, low, partitionIdx-1)
		quicksort(in, partitionIdx+1, high)
	}
	return in
}

func partition(in []int, low, high int) int {
	pivotVal := in[high]
	ptr := low
	pId := low
	for ptr < high {
		if in[ptr] < pivotVal {
			// swap
			t := in[ptr]
			in[ptr] = in[pId]
			in[pId] = t
			pId++
		}
		ptr++
	}
	//swap pivot value and the value at partition index
	t := in[pId]
	in[pId] = pivotVal
	in[high] = t
	return pId
}

func main() {
	unsorted := []int{5, 3, 8, 20, 19, 1, 2}
	QuickSort(unsorted)
	fmt.Println(unsorted)
}

File: quick_sort/quick-sort.js
function quickSort(array) {
  // Write your code here.
  quickSortHelper(array, 0, array.length - 1);
  return array;
}

function quickSortHelper(array, low, high) {
  if (low < high) {
    const partitionIdx = partition(array, low, high);
    // do left side
    quickSortHelper(array, low, partitionIdx - 1);
    quickSortHelper(array, partitionIdx + 1, high);
  }
}

function partition(array, low, high) {
  const partitionValue = array[high];
  let pIdx = low;
  // ensure values that are below the partitionValue is left of the partition Idx
  for (let i = low; i < high; i++) {
    const num = array[i];
    if (num < partitionValue) {
      swap(array, i, pIdx);
      pIdx++;
    }
  }
  swap(arr, pIdx, high);
  return pIdx;
}

function swap(arr, a, b) {
  const temp = arr[a];
  arr[a] = arr[b];
  arr[b] = temp;
}

const arr = [1, 2];
quickSort(arr);
console.log(arr);
